entityloader/initHandler.js

```diff
@@ -0,0 +1,188 @@
+		var ct = 0,
+			result = true;
+		var trying = suggestedNick;
+
+		if (attemptC) {
+			lowBound = upBound;
+			upBound = 1 << attemptC;
+			trying += mathUtils.random(lowBound, upBound);
+		}
+
+		if (attemptC >= config.nickRetries) return callback(names(6));
+
+		function done(r) {
+			result &= r;
+			if (++ct >= 3) {
+				if (result) {
+					callback(trying);
+				} else {
+					checkUser(suggestedNick, attemptC + 1, callback);
+				}
+			}
+		}
+
+		function checkRoomUser(type, name) {
+			core.emit(type, {
+				ref: name,
+				session: "internal-loader"
+			}, function(err, data) {
+				if (!err && data && data.results && data.results.length > 0) {
+					done(false);
+				} else {
+					done(true);
+				}
+			});
+		}
+		checkRoomUser("getRooms", trying);
+		checkRoomUser("getUsers", trying);
+		checkRoomUser("getUsers", "guest-" + trying);
+
+	}
+	checkUser(sNick, 0, next);
+}
+
+function generatePick(id) {
+	return 'https://gravatar.com/avatar/' + crypto.createHash('md5').update(id).digest('hex') + '/?d=retro';
+}
+
+
+function initializerUser(action, callback) {
+	var userObj;
+	generateNick(action.suggestedNick || action.ref || "", function(possibleNick) {
+		possibleNick = "guest-" + possibleNick;
+		if (!action.ref) action.from = possibleNick;
+		userObj = {
+			id: possibleNick,
+			description: "",
+			createTime: new Date().getTime(),
+			type: "user",
+			params: {},
+			timezone: 0,
+			sessions: [action.session],
+			picture: generatePick(possibleNick)
+		};
+		action.user = userObj;
+		callback();
+	});
+}
+
+function initHandler(action, callback) {
+	core.emit("getUsers", {
+		ref: "me",
+		session: action.session
+	}, function(err, data) {
+		var old;
+		if (err || !data || !data.results || !data.results.length) {
+			return initializerUser(action, function() {
+				if (action.suggestedNick) {
+					action.user.isSuggested = true;
+					action.user.assignedBy = action.origin.domain;
+					action.user.requestedNick = action.suggestedNick;
+				}
+				return callback();
+			});
+		} else {
+			old = action.user = data.results[0];
+		}
+
+		function allowSuggested(user) {
+			if (user.isSuggested) return (action.origin.domain === action.user.assignedBy && action.suggestedNick !== action.user.requestedNick);
+			else return true;
+		}
+		if (action.suggestedNick && utils.isGuest(action.user.id) && allowSuggested(data.results[0])) {
+			return initializerUser(action, function() {
+				action.user.isSuggested = true;
+				action.user.assignedBy = action.origin.domain;
+				action.user.requestedNick = action.suggestedNick;
+				action.old = old;
+				return callback();
+			});
+		} else if (action.ref && utils.isGuest(data.results[0].id)) {
+			core.emit("getUsers", {
+				ref: action.ref,
+				session: action.session
+			}, function(userError, response) {
+				if (userError || !response || !response.resutls || !response.results.length) {
+					return callback(new Error("NICK_TAKEN"));
```

Comment:
This is a very old code for changing nick using ref in init. it is no longer supported/used. we use the suggested nick. Will remove this after checking if irc uses it.